Lib/Hash.hpp

```diff
@@ -41,47 +63,81 @@ struct HashUtils
   { return combine(h1, combine(h2, h3, ts...)); }
 };
 
-template<class ElementHash>
-struct StackHash {
+// the identity hash
+// not a great idea in general: things usually distribute badly in some way
+// for example, pointers are usually aligned to e.g. multiples of 4
+// however, for e.g. variables (evenly distributed close to 0) it's very effective
+// also OK for secondary hashes
+struct IdentityHash
+{
   template<typename T>
-  static unsigned hash(const Stack<T>& s) {
-    unsigned res = 2166136261u;
-    typename Stack<T>::ConstIterator it(s);
-    while(it.hasNext()) {
-      res = HashUtils::combine(res, ElementHash::hash(it.next()));
-    }
-    return res;
+  static bool equals(T o1, T o2)
+  { return o1 == o2; }
+
+  template<typename T>
+  static unsigned hash(T val)
+  { return static_cast<unsigned>(val); }
+};
+
+// combine two identity hashes together
+// the implementation is historical and not known to work well
+// TODO: is this reasonable?
+struct PairIdentityHash {
+  template<typename T>
+  static unsigned hash(T pp) {
+    return static_cast<unsigned>(pp.first^pp.second^(pp.first<<1));
   }
 };
 
-template<class T>
+// wrapper around std::hash
```

Comment:
Some of the following custom hash structs had a `T` parameter that wasn't strictly necessary. I pushed it into the function as an overload instead which makes it a bit nicer to type...but it doesn't have to be that way if there's some reason not to.